File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/scala/org/rgn/akka/ch02/scala/mapreduce/actors/AggregateActor.scala
package org.rgn.akka.ch02.scala.mapreduce.actors

import akka.actor.Actor
import scala.collection.mutable.HashMap
import org.rgn.akka.ch02.scala.mapreduce.ReduceData
import org.rgn.akka.ch02.scala.mapreduce.Result

class AggregateActor extends Actor {

  val finalReduceMap = new HashMap[String, Int]

  def receive: Receive = {
    case ReduceData(reduceDataMap) =>
      aggregateInMemoryReduce(reduceDataMap)
    case Result =>
      sender ! finalReduceMap.toString
  }

  def aggregateInMemoryReduce(reduceList: Map[String, Int]): Unit = {
    for ((key, value) <- reduceList) {
      if (finalReduceMap contains key)
        finalReduceMap(key) = (value + finalReduceMap.get(key).get)
      else
        finalReduceMap += (key -> value)
    }
  }
}

File: rgn_concurrency/rgn_concurrency_akka_essentials/src/main/java/org/rgn/akka/ch02/mapreduce/actors/AggregateActor.java
package org.rgn.akka.ch02.mapreduce.actors;

import java.util.HashMap;
import java.util.Map;

import org.rgn.akka.ch02.mapreduce.messages.ReduceData;
import org.rgn.akka.ch02.mapreduce.messages.Result;

import akka.actor.UntypedActor;

public class AggregateActor extends UntypedActor {
	private Map<String, Integer> finalReduceMap = new HashMap<>();

	@Override
	public void onReceive(Object message) throws Exception {
		if (message instanceof ReduceData) {
			ReduceData reduceData = (ReduceData) message;
			aggregateInMemoryReduce(reduceData.getReduceList());
		} else if (message instanceof Result) {
			getSender().tell(finalReduceMap.toString(), getSelf());
		} else {
			unhandled(message);
		}
	}

	private void aggregateInMemoryReduce(Map<String, Integer> reduceList) {
		Integer count = null;
		for (String key : reduceList.keySet()) {
			if (finalReduceMap.containsKey(key)) {
				count = reduceList.get(key) + finalReduceMap.get(key);
				finalReduceMap.put(key, count);
			} else {
				finalReduceMap.put(key, reduceList.get(key));
			}
		}
	}

}
